### DISCUSSION TITLE
How to use pixi-based Docker image in Devcontainer?

### DISCUSSION THREAD

**Discussion by Robinlovelace** [prefix-dev/pixi#2088] [Q&A]
Here's the issue:

As described in #2082

```sh
docker run -it ghcr.io/geocompx/docker:r-pixi bash
R # run your R code!
```

works great. However, when I put that image in a Devcontainer I do not have access to R from the default terminal. Here's the Devcontainer: https://github.com/geocompx/geocompr/blob/079cc4dde61f9614f2acf8bce08ae3326cd1a82e/.devcontainer.json

Heres the Docker image: https://github.com/geocompx/docker/blob/master/pixi-r/Dockerfile

I wonder, is it something to do with the entrypoint?

**Comment by Robinlovelace** [comment]
See also https://github.com/prefix-dev/pixi-docker/issues/42

**Comment by ruben-arts** [comment]
I think it should help if you put the `shell-hook.sh` in a source command in the `~/.bashrc` this way the environment gets sourced on starting of the terminal. As it's basically just a activation script which sets some environment variables.

**Reply by Robinlovelace** [reply]
What about containers that use zsh or other? Maybe in a global .profile file would be better.

**Reply by ruben-arts** [reply]
That would be better, just make sure it's sourced and not run as a bash script separately.

**Comment by Robinlovelace** [comment]
Good news: it works! For some reason you must add the following postCreateCommand for it to work. This took many tries to get working. Any ideas why, and any ideas how to fix it upstream so that this is not needed in future?

I suspect a 'install global` option that installs pixi.toml globally would do it.

```
{
  "image": "ghcr.io/geocompx/docker:pixi-r",
  "customizations": {
    "vscode": {
      "extensions": ["reditorsupport.r"]
    }
  },
  "postCreateCommand": "apt update && apt install -y --no-install-recommends locales; echo \"en_US.UTF-8 UTF-8\" >> /etc/locale.gen; locale-gen",
}
```

See https://github.com/geocompx/geocompr/blob/77dd6207c3b0e1a51d62202c62d5d045c2961038/.devcontainer.json#L1-L9

**Reply by wolfv** [reply]
I guess that `apt` package could also be installed inside the docker image?

**Reply by Robinlovelace** [reply]
Will try... Travelling ATM, prs to the geocompx/docker repo welcome

**Reply by Robinlovelace** [reply]
@wolfv update on this: have tried your suggestion but it does not work. Moving command from 

```
  "postCreateCommand": "apt update && apt install -y --no-install-recommends locales; echo \"en_US.UTF-8 UTF-8\" >> /etc/locale.gen; locale-gen",
```

to 

```
RUN apt update && apt install -y --no-install-recommends locales; echo \"en_US.UTF-8 UTF-8\" >> /etc/locale.gen; locale-gen
```

In the Dockerfile leads to the same rather dramatic error shown below from inside the devcontainer: 

```
root@2651df13c344:/workspaces/geocompr# R

R version 4.4.1 (2024-06-14) -- "Race for Your Life"
Copyright (C) 2024 The R Foundation for Statistical Computing
Platform: x86_64-conda-linux-gnu

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

During startup - Warning messages:
1: Setting LC_CTYPE failed, using "C" 
2: Setting LC_COLLATE failed, using "C" 
3: Setting LC_TIME failed, using "C" 
4: Setting LC_MESSAGES failed, using "C" 
5: Setting LC_MONETARY failed, using "C" 
6: Setting LC_PAPER failed, using "C" 
7: Setting LC_MEASUREMENT failed, using "C" 

 *** caught segfault ***
address (nil), cause 'memory not mapped'

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Segmentation fault (core dumped)
```

This feels like a bug to me so will cross-reference from https://github.com/prefix-dev/pixi-docker/issues/42

Any ideas of what we can do in the Docker container to make it work?

For completeness, here are the exact versions of the Dockerfile and .devcontainer.json in which these tests were undertaken, not a minimal example but pretty small and self-contained:

- Dockerfile: https://github.com/geocompx/docker/blob/e22aef8f610874ad33ccc1b32504602b7bd0881d/pixi-r/Dockerfile#L1-L18
- .devcontainer.json: https://github.com/geocompx/geocompr/blob/77dd6207c3b0e1a51d62202c62d5d045c2961038/.devcontainer.json#L1-L9

**Reply by wolfv** [reply]
Maybe something like this works? https://stackoverflow.com/a/28406007
Sorry, I am also clueless here :)

**Reply by Robinlovelace** [reply]
Also came across that..

Solution: global install. See the pixi images here for currently solution, any ideas welcome. FYI: impressed at how quick it was to put these together. Any ideas how to avoid copy stages like this?

```
RUN ln -s /root/.pixi/envs/r/bin/* /usr/local/bin
```

Should the binaries already be on the path? Sure there's room for improvement here. Source: https://github.com/geocompx/docker/blob/master/pixi-r/Dockerfile

**Reply by wolfv** [reply]
Ah ok, I had a look at your Dockerfile. I think that's not the way we would encourage pixi usage :) The remaining issue when you went with `pixi.toml` was just about the locales?

**Reply by Robinlovelace** [reply]
Yes, as far as I can tell, the locales issues meant the R devcontainer only worked when postCreateCommand was set in the devcontainer for some reasons I don't understand. I think it works be great to identify the root cause of the locales issues and fix them upstream. Any guidance on best practices welcome, my take is that locales should 'just work', happy to do some more tests but not immediate thoughts on this and keen to hear what others think.
